OpenRA.Mods.Common/Traits/Modifiers/FrozenUnderFog.cs

```diff
@@ -87,19 +93,21 @@ public void Tick(Actor self)
 			VisibilityHash = 0;
 			foreach (var player in self.World.Players)
 			{
-				bool isVisible;
 				FrozenActor frozenActor;
+				bool isVisible;
 				if (!initialized)
 				{
-					frozen[player] = frozenActor = new FrozenActor(self, footprint, player.Shroud);
-					frozen[player].NeedRenderables = frozenActor.NeedRenderables = startsRevealed;
+					frozenActor = new FrozenActor(self, footprint, player.Shroud, startsRevealed);
+					isVisible = startsRevealed;
+					stateByPlayer.Add(player, new FrozenState(frozenActor) { IsVisible = isVisible });
 					player.PlayerActor.Trait<FrozenActorLayer>().Add(frozenActor);
```

Comment:
If you wanted to squeeze some more perf out of this, you could store this in another dictionary to save yourself some more trait lookups per tick.